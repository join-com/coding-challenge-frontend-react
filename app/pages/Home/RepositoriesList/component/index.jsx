import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import List from '../../../../atoms/List';
import LoadingIndicator from '../../../../molecules/LoadingIndicator';

import Repository from '../Repository';

import messages from './messages';
import styles from './styles.scss';

function RepositoriesList({ loading, error, repos }) {
  if(loading || (error !== false) || (repos !== false)) {
    return (
      <div className={styles.wrapper}>
        {loading ? <LoadingIndicator /> : <List className={styles.list} items={getListItems(repos)} />}
      </div>
    );
  }

  return null;
}

function getListItems(repos) {
  if(repos !== false) {
    return repos.map((repository) => <Repository repository={repository} />);
  }

  return [<FormattedMessage {...messages.somethingWrong} />];
}

RepositoriesList.propTypes = {
  loading: PropTypes.bool,
  error: PropTypes.any,
  repos: PropTypes.any,
};

export default RepositoriesList;
