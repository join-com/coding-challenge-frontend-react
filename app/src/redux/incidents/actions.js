import c from './constants';
import createAction from 'app/redux/createAction';

export const getIncidents = createAction({
    loadingType: c.INCIDENTS_LOADING,
    type: c.GET_INCIDENTS,
    apiName: 'getIncidents',
    defaultReq: {
        incident_type: 'theft',
        proximity: 'Berlin',
        proximity_square: 100
    }
});

export const getIncident = createAction({
    loadingType: c.INCIDENT_LOADING,
    type: c.GET_INCIDENT,
    apiName: 'getIncident'
});

export const getBike = createAction({
    loadingType: c.BIKE_LOADING,
    type: c.GET_BIKE,
    apiName: 'getBike'
});
