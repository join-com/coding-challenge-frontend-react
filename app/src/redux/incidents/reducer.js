import c from './constants';

const defaultState = {
    incidents: {
        uuid: undefined,
        loading: false,
        data: []
    },
    incident: {
        uuid: undefined,
        loading: false,
        data: {}
    },
    bike: {
        uuid: undefined,
        loading: false,
        data: {}
    }
};

export default function commentsReducer(state = defaultState, { type, payload }) {
    switch (type) {
        case c.INCIDENTS_LOADING:
            return {
                ...state,
                incidents: {
                    ...state.incidents,
                    ...payload
                }
            };

        case c.GET_INCIDENTS: {
            const {
                incidents: data,
                ...rest
            } = payload;

            return {
                ...state,
                incidents: {
                    data,
                    ...rest
                }
            };
        }

        case c.INCIDENT_LOADING:
            return {
                ...state,
                incident: {
                    ...state.incident,
                    ...payload
                }
            };

        case c.GET_INCIDENT: {
            const {
                incident: data,
                ...rest
            } = payload;

            return {
                ...state,
                incident: {
                    data,
                    ...rest
                }
            };
        }

        case c.BIKE_LOADING:
            return {
                ...state,
                bike: {
                    ...state.bike,
                    ...payload
                }
            };

        case c.GET_BIKE: {
            const {
                bikes: data,
                ...rest
            } = payload;

            return {
                ...state,
                bike: {
                    data,
                    ...rest
                }
            };
        }

        default:
            return state;
    }
}
