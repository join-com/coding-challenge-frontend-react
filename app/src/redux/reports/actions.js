import c from './constants';
import api from 'app/redux/api';

const defaultReq = {
    incident_type: 'theft',
    proximity: 'Berlin',
    proximity_square: 100
};

function setLoading(loading = false) {
    return {
        type: c.LOADING,
        payload: { loading }
    };
}

export function getIncidents(data = {}) {
    return dispatch => {
        dispatch(setLoading(true));

        (async () => {
            let payload;

            try {
                const res = await api.getIncidents({
                    ...defaultReq,
                    ...data
                });

                payload = {
                    status: 'success',
                    ...res.data
                };
            } catch (error) {
                payload = {
                    status: 'error',
                    error
                };
            }

            dispatch({
                type: c.GET_INCIDENTS,
                payload: {
                    uuid: Date.now(),
                    ...setLoading(false).payload,
                    ...payload
                }
            });
        })();
    };
}
