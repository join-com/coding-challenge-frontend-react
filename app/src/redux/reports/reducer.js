import c from './constants';

const defaultState = {
    uuid: undefined,
    incidents: [],
    loading: false
};

export default function commentsReducer(state = defaultState, { type, payload }) {
    switch (type) {
        case c.LOADING:
        case c.GET_INCIDENTS:
            return {
                ...state,
                ...payload
            };

        default:
            return state;
    }
}
