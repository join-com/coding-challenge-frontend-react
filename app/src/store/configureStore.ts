import { applyMiddleware, createStore } from 'redux'
import { createEpicMiddleware } from 'redux-observable'

import { reducer, epics } from '../redux'

export function configureStore() {
  const epicMiddleware = createEpicMiddleware()
  const store = createStore(reducer, applyMiddleware(epicMiddleware))
  epicMiddleware.run(epics)
  return store
}
