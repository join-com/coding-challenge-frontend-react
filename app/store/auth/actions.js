import {
  SIGN_IN_REQUEST,
  SIGN_IN_SUCCESS,
  SIGN_OUT,
  SIGN_UP_REQUEST,
} from './constants';

export function submitSignInRequest(data, intl, messages) {
  return {
    type: SIGN_IN_REQUEST,
    data,
    intl,
    messages,
  };
}

export function setSignInData(user, token) {
  return {
    type: SIGN_IN_SUCCESS,
    user,
    token,
  };
}

export function signOut() {
  return {
    type: SIGN_OUT,
  };
}

export function submitSignUpRequest(data, intl, messages) {
  return {
    type: SIGN_UP_REQUEST,
    data,
    intl,
    messages,
  };
}
