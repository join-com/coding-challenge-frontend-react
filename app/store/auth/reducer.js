import { fromJS } from 'immutable';

import {
  SIGN_IN_SUCCESS,
  SIGN_OUT,
} from './constants';

const initialState = fromJS({
  user: null,
});

function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case SIGN_IN_SUCCESS:
      return state.set('user', action.user);
    case SIGN_OUT:
      return state.set('user', null);
    default:
      return state;
  }
}

export default reducer;
