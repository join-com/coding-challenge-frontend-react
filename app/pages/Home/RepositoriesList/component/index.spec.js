import React from 'react';
import renderer from 'react-test-renderer';

import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';

import configureStore from '../../../../utils/store';
import { setRepositories } from '../../../../store/github/actions';
import reducer from '../../../../store/github/reducer';

import RepositoriesList from './';

describe('<RepositoriesList />', () => {
  it('renders a loading spinner if the loading flag is set', () => {
    expect(renderer.create(<RepositoriesList loading />).toJSON()).toMatchSnapshot();
  });

  it('renders an error message if an error occurred', () => {
    expect(
      renderer.create(
        <IntlProvider locale="en">
          <RepositoriesList error repos={false} />
        </IntlProvider>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it('renders nothing if the repository list is empty', () => {
    expect(renderer.create(<RepositoriesList error={false} repos={false} />).toJSON()).toMatchSnapshot();
  });

  it('renders the repository list if the list is not empty', () => {
    const repositories = [
      {
        id: 52143857,
        name: 'ta-maxlength',
        full_name: 'alexander-elgin/ta-maxlength',
        owner: {
          login: 'alexander-elgin',
        },
        html_url: 'https://github.com/alexander-elgin/ta-maxlength',
        open_issues_count: 3,
      },
      {
        id: 33667892,
        name: 'meetermaid',
        full_name: 'rasp3175/meetermaid',
        owner: {
          login: 'rasp3175',
        },
        html_url: 'https://github.com/rasp3175/meetermaid',
        open_issues_count: 0,
      },
    ];

    const currentUser = 'alexander-elgin';
    const store = configureStore({
      github: reducer,
    });
    store.dispatch(setRepositories(repositories, currentUser));

    expect(renderer.create(
      <Provider store={store}>
        <IntlProvider locale="en">
          <RepositoriesList error={false} repos={repositories} />
        </IntlProvider>
      </Provider>
    ).toJSON()).toMatchSnapshot();
  });
});
