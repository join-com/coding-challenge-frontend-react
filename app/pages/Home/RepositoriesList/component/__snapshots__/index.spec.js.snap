// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepositoriesList /> renders a loading spinner if the loading flag is set 1`] = `
<div
  className="wrapper"
>
  <div
    className="component__Wrapper-sc-1v3e81z-0 gjQjkD"
  >
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 lmtFYx"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 ilfvYA"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 dufGFp"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 eYmTeu"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 ksuiKX"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 klXKrK"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 kEwzjM"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 bbTZeY"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 MGAPO"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 jiLNcn"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 jAmdJg"
    />
    <div
      className="component__CirclePrimitive-sc-1ls99zz-0 Xlsli"
    />
  </div>
</div>
`;

exports[`<RepositoriesList /> renders an error message if an error occurred 1`] = `
<div
  className="wrapper"
>
  <ul
    className="list"
  >
    <li>
      <span>
        Something went wrong, please try again!
      </span>
    </li>
  </ul>
</div>
`;

exports[`<RepositoriesList /> renders nothing if the repository list is empty 1`] = `null`;

exports[`<RepositoriesList /> renders the repository list if the list is not empty 1`] = `
<div
  className="wrapper"
>
  <ul
    className="list"
  >
    <li>
      <div
        className="wrapper"
      >
        <a
          className="component__A-sc-1qxsdu7-0 component__RepoLink-sc-2iw2dp-0 cbLOVz"
          href="https://github.com/alexander-elgin/ta-maxlength"
          target="_blank"
        >
          ta-maxlength
        </a>
        <a
          className="component__A-sc-1qxsdu7-0 component__IssuesLink-sc-1rshot1-0 ezZxTa"
          href="https://github.com/alexander-elgin/ta-maxlength/issues"
          target="_blank"
        >
          <svg
            className="icon"
            height="1em"
            width="0.875em"
          >
            <path
              d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7S10.14 13.7 7 13.7 1.3 11.14 1.3 8s2.56-5.7 5.7-5.7m0-1.3C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7S10.86 1 7 1z m1 3H6v5h2V4z m0 6H6v2h2V10z"
            />
          </svg>
          <span>
            3
          </span>
        </a>
      </div>
    </li>
    <li>
      <div
        className="wrapper"
      >
        <a
          className="component__A-sc-1qxsdu7-0 component__RepoLink-sc-2iw2dp-0 cbLOVz"
          href="https://github.com/rasp3175/meetermaid"
          target="_blank"
        >
          rasp3175/meetermaid
        </a>
        <a
          className="component__A-sc-1qxsdu7-0 component__IssuesLink-sc-1rshot1-0 ezZxTa"
          href="https://github.com/rasp3175/meetermaid/issues"
          target="_blank"
        >
          <svg
            className="icon"
            height="1em"
            width="0.875em"
          >
            <path
              d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7S10.14 13.7 7 13.7 1.3 11.14 1.3 8s2.56-5.7 5.7-5.7m0-1.3C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7S10.86 1 7 1z m1 3H6v5h2V4z m0 6H6v2h2V10z"
            />
          </svg>
          <span>
            0
          </span>
        </a>
      </div>
    </li>
  </ul>
</div>
`;
